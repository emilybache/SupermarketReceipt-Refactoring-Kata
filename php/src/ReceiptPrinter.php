<?php

declare(strict_types=1);

namespace Supermarket;

use Supermarket\Model\{
    Discount,
    ProductUnit,
    Receipt,
    ReceiptItem
};

class ReceiptPrinter
{
    /**
     * @var int
     */
    private $columns;

    public function __construct(int $columns = 40)
    {
        $this->columns = $columns;
    }

    public function printReceipt(Receipt $receipt): string
    {
        $result = '';
        foreach ($receipt->getItems() as $item) {
            $itemPresentation = $this->presentReceiptItem($item);
            $result .= $itemPresentation;
        }

        foreach ($receipt->getDiscounts() as $discount) {
            $discountPresentation = $this->presentDiscount($discount);
            $result .= $discountPresentation;
        }

        $result .= "\n";
        $result .= $this->presentTotal($receipt);
        return $result;
    }

    protected function presentReceiptItem(ReceiptItem $item): string
    {
        $price = self::presentPrice($item->getTotalPrice());
        $name = $item->getProduct()->getName();

        $line = $this->formatLineWithWhitespace($name, $price) . "\n";

        if ($item->getQuantity() !== 1.0) {
            $line .= '  ' . self::presentPrice($item->getPrice()) . ' * ' . self::presentQuantity($item) . "\n";
        }
        return $line;
    }

    protected function presentDiscount(Discount $discount): string
    {
        $name = "{$discount->getDescription()}({$discount->getProduct()->getName()})";
        $value = self::presentPrice($discount->getDiscountAmount());

        return $this->formatLineWithWhitespace($name, $value) . "\n";
    }

    protected function presentTotal(Receipt $receipt): string
    {
        $name = 'Total: ';
        $value = self::presentPrice($receipt->getTotalPrice());
        return $this->formatLineWithWhitespace($name, $value);
    }

    protected function formatLineWithWhitespace(string $name, string $value): string
    {
        $whitespaceSize = $this->columns - strlen($name) - strlen($value);
        return $name . str_repeat(' ', $whitespaceSize) . $value;
    }

    protected static function presentPrice(float $price): string
    {
        return sprintf('%.2F', $price);
    }

    private static function presentQuantity(ReceiptItem $item): string
    {
        return $item->getProduct()->getUnit()->equals(ProductUnit::EACH()) ?
            sprintf('%x', $item->getQuantity()) :
            sprintf('%.3F', $item->getQuantity());
    }
}
