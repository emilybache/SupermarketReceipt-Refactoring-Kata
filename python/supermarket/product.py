from dataclasses import dataclass
from enum import Enum


ProductName = str


class ProductUnit(Enum):
    EACH = 1
    KILO = 2


@dataclass(frozen=True)
class Product:
    name: ProductName
    unit: ProductUnit


@dataclass(frozen=True)
class ProductQuantity:
    product: Product
    quantity: float
