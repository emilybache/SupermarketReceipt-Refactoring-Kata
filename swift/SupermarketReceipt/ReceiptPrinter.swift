public class ReceiptPrinter {

    private var columns: Int = 40

    public init(columns: Int) {
        self.columns = columns
    }

    public func printReceipt(receipt: Receipt) -> String {
        var result = ""
        for item in receipt.items {
            var price = String(format: "%.2f", item.totalPrice)
            var quantity = ReceiptPrinter.presentQuantity(item: item)
            var name = item.product.name
            var unitPrice = String(format :"%.2f", item.price)

            var whitespaceSize = self.columns - name.count - price.count
            var line = name + ReceiptPrinter.getWhitespace(whitespaceSize: whitespaceSize) + price + "\n"

            if (item.quantity != 1) {
                line += "  " + unitPrice + " * " + quantity + "\n"
            }
            result.append(line)
        }
        for discount in receipt.discounts {
            var productPresentation = discount.product.name
            var pricePresentation = String(format: "%.2f", discount.discountAmount)
            var description = discount.description
            result.append(description)
            result.append("(")
            result.append(productPresentation)
            result.append(")")
            result.append(ReceiptPrinter.getWhitespace(whitespaceSize: self.columns - 3 - productPresentation.count - description.count - pricePresentation.count))
            result.append("-")
            result.append(pricePresentation)
            result.append("\n")
        }
        result.append("\n")
        var pricePresentation = String(format: "%.2f", Double(receipt.getTotalPrice()))
        var total = "Total: "
        var whitespace = ReceiptPrinter.getWhitespace(whitespaceSize: self.columns - total.count - pricePresentation.count)
        result.append(total)
        result.append(whitespace)
        result.append(pricePresentation)
        return result
    }

    private static func presentQuantity(item: ReceiptItem ) -> String {
        return ProductUnit.Each == item.product.unit
            ? String(format: "%x", Int(item.quantity))
            : String(format: "%.3f", item.quantity)
    }

    private static func getWhitespace(whitespaceSize: Int) -> String {
        var whitespace = ""
        for i in 0..<whitespaceSize {
            whitespace.append(" ")
        }
        return whitespace
    }
}
