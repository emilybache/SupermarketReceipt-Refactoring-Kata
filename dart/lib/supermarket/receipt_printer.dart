import 'package:supermarket_receipt/supermarket/model/product_unit.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/receipt_item.dart';
import 'package:supermarket_receipt/supermarket/model/discount.dart';

class ReceiptPrinter {
  
  final int _columns;

  ReceiptPrinter() : this._columns = 40;

  ReceiptPrinter.withColumns(this._columns);

  String printReceipt(Receipt receipt) {
    StringBuffer result = StringBuffer();
    for (ReceiptItem item in receipt.getItems()) {
      String receiptItem = presentReceiptItem(item);
      result.write(receiptItem);
    }
    for (Discount discount in receipt.getDiscounts()) {
      String discountPresentation = presentDiscount(discount);
      result.write(discountPresentation);
    }

    result.write('\n');
    result.write(presentTotal(receipt));
    return result.toString();
  }

  String presentReceiptItem(ReceiptItem item) {
    String totalPricePresentation = presentPrice(item.getTotalPrice());
    String name = item.getProduct().getName();

    String line = formatLineWithWhitespace(name, totalPricePresentation);

    if (item.getQuantity() != 1) {
      line += '  ' + presentPrice(item.getPrice()) + ' * ' + presentQuantity(item) + '\n';
    }
    return line;
  }

  String presentDiscount(Discount discount) {
    String name = discount.getDescription() + ' (' + discount.getProduct().getName() + ')';
    String value = presentPrice(discount.getDiscountAmount());

    return formatLineWithWhitespace(name, value);
  }

  String presentTotal(Receipt receipt) {
    String name = 'Total: ';
    String value = presentPrice(receipt.getTotalPrice());
    return formatLineWithWhitespace(name, value);
  }

  String formatLineWithWhitespace(String name, String value) {
    StringBuffer line = StringBuffer();
    line.write(name);
    int whitespaceSize = _columns - name.length - value.length;
    for (int i = 0; i < whitespaceSize; i++) {
      line.write(' ');
    }
    line.write(value);
    line.write('\n');
    return line.toString();
  }

  String presentPrice(double price) {
    return price.toStringAsFixed(2);
  }

  String presentQuantity(ReceiptItem item) {
    if (item.getProduct().getUnit() == ProductUnit.EACH) {
      return item.getQuantity().toInt().toString();
    } else {
      return item.getQuantity().toStringAsFixed(3);
    }
  }
}