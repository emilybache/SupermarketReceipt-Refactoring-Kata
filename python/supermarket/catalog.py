from typing import Dict

from supermarket.product import Product, ProductName


class SupermarketCatalog:
    def __init__(self) -> None:
        self.products: Dict[ProductName, Product] = {}
        self.prices: Dict[ProductName, float] = {}

    def add_product(self, product: Product, price: float) -> None:
        raise Exception("cannot be called from a unit test - it accesses the database")

    def unit_price(self, product_name: ProductName) -> float:
        raise Exception("cannot be called from a unit test - it accesses the database")
