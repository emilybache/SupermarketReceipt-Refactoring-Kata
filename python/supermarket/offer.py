import math
from abc import ABCMeta, abstractmethod
from enum import Enum
from dataclasses import dataclass
from typing import Any, Optional, Sequence

from supermarket.product import Product


@dataclass(frozen=True)
class Discount:
    product: Product
    description: str
    discount_amount: float


class SpecialOfferType(Enum):
    PERCENTAGE_DISCOUNT = 1
    QUANTITY_FOR_AMOUNT = 2
    QUANTITY_FOR_QUANTITY = 3


class Offer(metaclass=ABCMeta):
    def __init__(self, product: Product, argument: Optional[Any] = None) -> None:
        self.product = product
        self.argument = argument

    def calculate_discount(self, unit_price: float, purchased_quantity: float):
        raise NotImplementedError


class QuantityForQuantity(Offer):

    TYPE = SpecialOfferType.QUANTITY_FOR_QUANTITY

    def calculate_discount(self, unit_price: float, purchased_quantity: float):
        assert isinstance(self.argument, Sequence)
        quantity_x, quantity_y = self.argument
        purchased_quantity_as_int = int(purchased_quantity)
        number_of_x = math.floor(purchased_quantity_as_int / quantity_x)
        if purchased_quantity_as_int > quantity_y:
            discount_amount = purchased_quantity * unit_price - (
                (number_of_x * quantity_y * unit_price) + purchased_quantity_as_int % quantity_x * unit_price
            )
            return Discount(
                self.product,
                f"{quantity_x} for {quantity_y}",
                -discount_amount
            )


class QuantityForAmount(Offer):

    TYPE = SpecialOfferType.QUANTITY_FOR_AMOUNT

    def calculate_discount(self, unit_price: float, purchased_quantity: float):
        assert isinstance(self.argument, Sequence)
        offer_quantity, offer_amount = self.argument
        if purchased_quantity < offer_quantity:
            return

        purchased_quantity_as_int = int(purchased_quantity)
        effective_offers = math.floor(purchased_quantity_as_int / offer_quantity)
        remaining_quantity = purchased_quantity_as_int % offer_quantity
        total_amount = (
            offer_amount * effective_offers + remaining_quantity * unit_price
        )
        discount_amount = unit_price * purchased_quantity - total_amount
        discount_text = f"{offer_quantity} for {offer_amount}"
        return Discount(self.product, discount_text, -discount_amount)


class PercentageDiscount(Offer):

    TYPE = SpecialOfferType.PERCENTAGE_DISCOUNT

    def calculate_discount(self, unit_price: float, purchased_quantity: float):
        discount_percent = self.argument
        assert isinstance(discount_percent, float)
        return Discount(
            self.product,
            f"{self.argument}% off",
            -purchased_quantity * unit_price * discount_percent / 100.0,
        )


SPECIAL_OFFER_IMPLEMENTATIONS = {
    SpecialOfferType.QUANTITY_FOR_QUANTITY: QuantityForQuantity,
    SpecialOfferType.PERCENTAGE_DISCOUNT: PercentageDiscount,
    SpecialOfferType.QUANTITY_FOR_AMOUNT: QuantityForAmount,
}


def create_offer(
    offer_type: SpecialOfferType, product: Product, argument: Optional[Any] = None
):
    try:
        return SPECIAL_OFFER_IMPLEMENTATIONS[offer_type](product, argument)
    except KeyError:
        raise NotImplementedError(f"Unknown offer type: {offer_type}")
