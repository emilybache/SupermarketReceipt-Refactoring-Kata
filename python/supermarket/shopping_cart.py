from supermarket.product import ProductName, ProductQuantity, Product

from collections import defaultdict
from typing import List, Dict


class ShoppingCart:
    def __init__(self) -> None:
        self._items: List[ProductQuantity] = []
        self._product_quantities: Dict[ProductName, float] = defaultdict(float)

    @property
    def items(self) -> List[ProductQuantity]:
        return self._items

    def add_item(self, product: Product) -> None:
        self.add_item_quantity(product, 1.0)

    @property
    def product_quantities(self) -> Dict[ProductName, float]:
        return self._product_quantities

    def add_item_quantity(self, product: Product, quantity: float) -> None:
        self._items.append(ProductQuantity(product, quantity))
        self._product_quantities[product.name] += quantity
