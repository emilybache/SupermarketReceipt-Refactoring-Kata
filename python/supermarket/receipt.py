from dataclasses import dataclass
from typing import List

from supermarket.product import Product
from supermarket.offer import Discount


@dataclass(frozen=True)
class ReceiptItem:
    product: Product
    quantity: float
    price: float
    total_price: float


class Receipt:
    def __init__(self) -> None:
        self._items: List[ReceiptItem] = []
        self._discounts: List[Discount] = []

    def total_price(self) -> float:
        return sum(i.total_price for i in self.items) + sum(
            d.discount_amount for d in self.discounts
        )

    def add_product(
        self, product: Product, quantity: float, price: float, total_price: float
    ) -> None:
        # TODO: pass the object instead of so many arguments
        self._items.append(ReceiptItem(product, quantity, price, total_price))

    def add_discount(self, discount: Discount) -> None:
        self._discounts.append(discount)

    @property
    def items(self) -> List[ReceiptItem]:
        return self._items[:]

    @property
    def discounts(self) -> List[Discount]:
        return self._discounts[:]
