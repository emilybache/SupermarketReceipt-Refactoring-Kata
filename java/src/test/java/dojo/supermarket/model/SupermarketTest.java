package dojo.supermarket.model;

import dojo.supermarket.ReceiptPrinter;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class SupermarketTest {
    private SupermarketCatalog catalog;
    private Teller teller;
    private ShoppingCart theCart;
    private Product toothbrush;
    private Product rice;
    private Product apples;
    private Product cherryTomatoes;

    @BeforeEach
    public void setUp() {
        catalog = new FakeCatalog();
        teller = new Teller(catalog);
        theCart = new ShoppingCart();

        toothbrush = new Product("toothbrush", ProductUnit.Each);
        catalog.addProduct(toothbrush, 0.99);
        rice = new Product("rice", ProductUnit.Each);
        catalog.addProduct(rice, 2.99);
        apples = new Product("apples", ProductUnit.Kilo);
        catalog.addProduct(apples, 1.99);
        cherryTomatoes = new Product("cherry tomato box", ProductUnit.Each);
        catalog.addProduct(cherryTomatoes, 0.69);

    }

    @Test
    public void an_empty_shopping_cart_should_cost_nothing() {
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void one_normal_item() {
        theCart.addItem(toothbrush);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void two_normal_items() {
        theCart.addItem(toothbrush);
        theCart.addItem(rice);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void buy_two_get_one_free() {
        theCart.addItem(toothbrush);
        theCart.addItem(toothbrush);
        theCart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, catalog.getUnitPrice(toothbrush));
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void buy_two_get_one_free_but_insufficient_in_basket() {
        theCart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, catalog.getUnitPrice(toothbrush));
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }
    @Test
    public void buy_five_get_one_free() {
        theCart.addItem(toothbrush);
        theCart.addItem(toothbrush);
        theCart.addItem(toothbrush);
        theCart.addItem(toothbrush);
        theCart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, catalog.getUnitPrice(toothbrush));
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void loose_weight_product() {
        theCart.addItemQuantity(apples, .5);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void percent_discount() {
        theCart.addItem(rice);
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, rice, 10.0);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void xForY_discount() {
        theCart.addItem(cherryTomatoes);
        theCart.addItem(cherryTomatoes);
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, cherryTomatoes,.99);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void xForY_discount_with_insufficient_in_basket() {
        theCart.addItem(cherryTomatoes);
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, cherryTomatoes,.99);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount() {
        theCart.addItemQuantity(apples, 5);
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples,6.99);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount_withSix() {
        theCart.addItemQuantity(apples, 6);
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples,5.99);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount_withSixteen() {
        theCart.addItemQuantity(apples, 16);
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples,7.99);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount_withFour() {
        theCart.addItemQuantity(apples, 4);
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples,8.99);
        Receipt receipt = teller.checksOutArticlesFrom(theCart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }
}
