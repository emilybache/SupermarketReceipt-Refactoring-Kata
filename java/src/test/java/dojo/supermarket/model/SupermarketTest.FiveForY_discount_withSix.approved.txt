apples                             11.94
  1.99 * 6.000
5 for 5.99(apples)                 -3.96

Total:                              7.98